{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Usage Tutorial (EBV vs Others):\n",
    "\n",
    "This tutorial uses the EBV vs Others (TCGA) task as an example to describe. It is a two-classification task (can be extended to multi-classification tasks), including EBV and Others. \n",
    "\n",
    "## Install Environment:\n",
    "\n",
    "Create the environment with conda commands:\n",
    "```\n",
    "conda create -n conceppath python=3.8.18\n",
    "conda activate conceppath\n",
    "```\n",
    "\n",
    "Install the dependencies:\n",
    "```\n",
    "git clone https://github.com/HKU-MedAI/ConcepPath.git\n",
    "cd ConcepPath\n",
    "pip install -r requirements.txt\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The [TCGA data](https://portal.gdc.cancer.gov/) can be accessed from this link. And the processing code mainly refers to [CLAM](https://github.com/mahmoodlab/CLAM).\n",
    "\n",
    "The folder structure of the this part: ({xxx} represents variables)\n",
    "\n",
    "```\n",
    "└── {experiment_root_path}\n",
    "    ├── data\n",
    "        ├── datasets\n",
    "            └── {dataset_name}\n",
    "                └── {task_name}_label.csv\n",
    "        └── raw_data\n",
    "            └── {dataset_name}\n",
    "                ├── csv\n",
    "                    ├── bm.csv\n",
    "                    └── pl_mag{target_mag}x_patch{base_patch_size}_{target_patch_size}.csv\n",
    "                └── segmented_patch\n",
    "                    └── pl_mag{target_mag}x_patch{base_patch_size}_{target_patch_size}\n",
    "                        ├── masks\n",
    "                            └── {wsi_name}.jpg\n",
    "                        ├── patches\n",
    "                            └── {wsi_name}.h5\n",
    "                        ├── stitches\n",
    "                            └── {wsi_name}.jpg\n",
    "                        └── process_list_autogen.csv\n",
    "                        \n",
    "    └── {experiment_name}\n",
    "        ├── input\n",
    "            ├── csv\n",
    "                ├── split\n",
    "                    └── {fold_i}.csv\n",
    "            └── prompt\n",
    "                ├── patch_prompts.json\n",
    "                └── region_prompts.json\n",
    "        └── output\n",
    "            ├── attn_score\n",
    "                └──{vlm_name_i}\n",
    "                    └── {wsi_name}_{vlm_name}.pkl\n",
    "            ├── heatmap\n",
    "                └── attn_map\n",
    "                    └── {vlm_name}\n",
    "                        └── {wsi_name}\n",
    "                            ├── aa_thumbnail.png\n",
    "                            └── {patch_prompt_i}.png\n",
    "            ├── metrics\n",
    "                └── metrics.csv\n",
    "            └── model\n",
    "                └── {model_name}\n",
    "\n",
    "```\n",
    "\n",
    "Steps include：      \n",
    "1. Descriptive Prompt Generation;\n",
    "2. Image Data Preprocessing;\n",
    "3. Training;\n",
    "4. Testing and saving attention scores;\n",
    "5. Postprocessing;\n",
    "6. Other"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Descriptive Prompt Generation:\n",
    "\n",
    "Generate descriptive prompt on two levels: slide-level and patch-level. Templates for questioning on LLM (GPT-4) are:\n",
    "Patch-level + Postive(EBV):\n",
    "```txt\n",
    "Q1: Please provide a summary of the factors found in primary tumor whole slide images that may indicate Epstein-Barr Virus (EBV)-positive subtype of Gastric cancer, along with a description of their image features in short terms, separated by semicolons. Please avoid using subtype names in your response.\n",
    "Q*: Please avoid using the word \"EBV\" in your response.\n",
    "Q*: Please give more.\n",
    "```\n",
    "Patch-level + Negative(others):\n",
    "```\n",
    "Q2: Suppose we let Gastric cancer subtypes other than Epstein-Barr Virus (EBV)-positive subtype into one group. Please provide a summary of the factors found in primary tumor whole slide images that may indicate this group of subtypes, along with a description of their image features in short terms, separated by semicolons.\n",
    "Q*: Please avoid using the word \"EBV\" in your response.\n",
    "Q*: Please give more.\n",
    "```\n",
    "Slide-level + Postive(EBV):\n",
    "```txt\n",
    "Q3: Summary the appearance of whole slide images of Epstein-Barr Virus (EBV)-positive subtype of Gastric cancer in short terms, separated by semicolons.\n",
    "Q*: Please avoid using the word \"EBV\" in your response.\n",
    "Q*: Please give more.\n",
    "```\n",
    "Slide-level + Negative(others):\n",
    "```\n",
    "Q4: Suppose we let Gastric cancer subtypes other than Epstein-Barr Virus (EBV)-positive subtype into one group. Summary the appearance of whole slide images of this group in short terms, separated by semicolons. \n",
    "Q*: Please avoid using the word \"EBV\" in your response.\n",
    "Q*: Please give more.\n",
    "```\n",
    "\n",
    "Write these answers with format as the following: (columns=\"prompt_level,label,descriptive_prompt\")\n",
    "```csv\n",
    "patch,ebv,Lymphoid stroma: Dense lymphoid stroma with infiltrating lymphocytes; lymphocytes surrounding tumor cells.\n",
    "patch,ebv,Lymphoepithelioma-like appearance: Undifferentiated tumor cells with abundant lymphocytic infiltrate; tumor cells interspersed with lymphocytes.\n",
    "patch,ebv,Epstein-Barr Virus (EBV)-encoded RNA (EBER) expression: Strong and diffuse nuclear staining for EBER; intense staining in the nuclei of tumor cells.\n",
    "...\n",
    "slide,ebv,Lymphocyte-rich infiltrate; Well-defined glandular structures; Epithelial cell apoptosis; Nuclear atypia; Syncytial growth pattern; Presence of viral-associated features; Intense lymphoid reaction; Absence of signet ring cells; Expression of viral RNA.\n",
    "```\n",
    "Next, run the helper script to parse these prompts into JSON format. The primary functions included within the helper scripts are:\n",
    " - split prompts into two file based on its level (slide vs.patch)\n",
    " - avoid using class label(or associated keywords) \n",
    " - filter the top-N concepts of patch-level prompts\n",
    "\n",
    "Note: `EXPERIMENT_NAME` should be the same as used in `main.py`.\n",
    "```bash\n",
    "EXPERIMENT_NAME=molecular_ebv_others_full_train python tools/parse_prompt.py\n",
    "```\n",
    "\n",
    "And the references used in this section listed in 6.1 Prompt References."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Image Data Preprocessing:\n",
    "\n",
    "## 2.1 Generate processing list:\n",
    "\n",
    "### Parameter Description:\n",
    "- **`--exp_root_path:`**: root path of experiment;\n",
    "- **`--WSI_dir:`**: WSIs' saving directory;\n",
    "- **`--save_dir`**: result's saving directory;\n",
    "- **`--base_patch_size`**: basic patch size;\n",
    "- **`--target_mag`**: target magnification;\n",
    "- **`--number`**: number of processing files;\n",
    "- **`--WSI_name`**: WSI dataset (/{save_dir}/{WSI_name})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils.processing_utils import *\n",
    "import os\n",
    "\n",
    "exp_root_path = \"/home/r10user13/ConcepPath\"                           \n",
    "\n",
    "generate_pl_bm(\n",
    "        WSI_dir=\"/data1/r10user13/TCGA-WSI/STAD/STAD\",\n",
    "        save_dir= os.path.join(exp_root_path, \"/data/raw_data/\"),\n",
    "        base_patch_size=448,\n",
    "        target_mag=20,\n",
    "        number=\"all\",\n",
    "        WSI_name=\"STAD\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.2 Segment patches and save coords information:\n",
    "\n",
    "### Parameter Description:\n",
    "- **`--source:`**: WSIs' saving directory;\n",
    "- **`--save_dir:`**: save directory;\n",
    "- **`--patch_size`**: patch size;\n",
    "- **`--step_size`**: step size (if it is less than patch size, there will be overlap, if it is greater than patch size, there will be gaps);\n",
    "- **`--seg`**: whether to generate mask;\n",
    "- **`--patch`**: whether to generate a patch;\n",
    "- **`--stitch`**: whether to generate stitch;\n",
    "- **`--process_list`**: process list;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python create_patches_fp.py \\\n",
    "    --source /data1/r10user3/TCGA-WSI/STAD/STAD \\                                               \n",
    "    --save_dir /home/r10user13/ConcepPath/data/raw_data/STAD/segmented_patch/pl_mag20x_patch448_448 \\\n",
    "    --patch_size 448 \\\n",
    "    --step_size 448 \\\n",
    "    --seg \\\n",
    "    --patch \\\n",
    "    --stitch \\\n",
    "    --process_list /home/r10user13/ConcepPath/data/raw_data/STAD/csv/pl_mag20x_patch448_448.csv"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.3 Integrate the dataset and generate label files:\n",
    "\n",
    "### Parameter Description:\n",
    "- **`--raw_label:`**: raw label saving path;\n",
    "- **`--dataset_path`**: result's saving path;\n",
    "- **`--file_name`**: saving file name;\n",
    "- **`--seg_patch_dir`**: path of segmented patches;\n",
    "- **`--WSI_dir`**: WSIs' saving directory;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob, os\n",
    "from utils.processing_utils import *\n",
    "\n",
    "raw_label = '/home/r10user13/ConcepPath/data/molecular_label_raw.csv'\n",
    "dataset_path = \"/home/r10user13/ConcepPath/data/datasets/STAD\"\n",
    "file_name = \"molecular_ebv_others.csv\"\n",
    "seg_patch_dir = '/home/r10user13/ConcepPath/data/raw_data/STAD/segmented_patch'\n",
    "WSI_dir = \"/data1/r10user3/TCGA-WSI/STAD/STAD\"\n",
    "\n",
    "file_list = glob.glob(os.path.join(WSI_dir, \"*\"))\n",
    "raw_label_df = pd.read_csv(raw_label)\n",
    "label_dict = {\n",
    "    \"slide_fp\": [],\n",
    "    \"label\": []\n",
    "}\n",
    "\n",
    "def panduan(label):\n",
    "    if label == \"EBV\":\n",
    "        return \"EBV\"\n",
    "    else:\n",
    "        return \"others\"\n",
    "    \n",
    "for i, row in raw_label_df[raw_label_df[\"Note\"].isna()].iterrows():\n",
    "    result = [s for s in file_list if all(sub in s for sub in [row['TCGA barcode']])]\n",
    "    label = panduan(row['Molecular Subtype'])\n",
    "    label_dict[\"slide_fp\"] += result\n",
    "    label_dict[\"label\"] += [label]*len(result)\n",
    "\n",
    "label_map = dict(zip(label_dict[\"slide_fp\"], label_dict[\"label\"]))\n",
    "\n",
    "directory_paths = {\n",
    "    seg_patch_dir: WSI_dir\n",
    "}\n",
    "    \n",
    "generate_label_file(\n",
    "    directory_paths, \n",
    "    dataset_path, \n",
    "    file_name, \n",
    "    label_map)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.4 Feature Extraction:\n",
    "\n",
    "### Parameter Description:\n",
    "- **`--vlm_model:`**: vision-language model name;\n",
    "- **`--label_fp:`**: path of label file;\n",
    "- **`--batch_size`**: batch size;\n",
    "- **`--save_rp`**: saving path of extracted features;\n",
    "- **`--num_workers`**: number of workers;\n",
    "- **`--base_mag`**: basic magnification;\n",
    "- **`--base_patch_size`**: basic patch size;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!python create_feature_extraction.py  \\\n",
    "    --vlm_model quilt1m \\\n",
    "    --label_fp /home/r10user13/ConcepPath/data/datasets/STAD/molecular_ebvmsi_others.csv  \\\n",
    "    --batch_size 128 \\\n",
    "    --save_rp /data2/r10user13/ConcepPath \\\n",
    "    --num_workers 32 \\\n",
    "    --base_mag 20 \\\n",
    "    --base_patch_size 448"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2.5 Dataset split:\n",
    "\n",
    "### Parameter Description:\n",
    "- **`--shot_num:`**: number of shot [1,2,4,8,16,32,64, ... ,full_train];\n",
    "- **`--exp_rp:`**: root path of experiment;\n",
    "- **`--exp_name`**: name of experiment;\n",
    "- **`--fold_num`**: number of fold;\n",
    "- **`--labe_fp`**: path of label file (generated in step);\n",
    "- **`--base_mag`**: basic magnification;\n",
    "- **`--base_patch_size`**: basic patch size;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# split\n",
    "from sklearn.model_selection import train_test_split, StratifiedKFold\n",
    "import os, shutil, random\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# shot_num = [1,2,4,8,16,32,64]\n",
    "shot_num = \"full_train\"\n",
    "exp_rp = \"/home/r10user13/ConcepPath/experiment\"\n",
    "exp_name = \"stad_molecular_ebv_others_ft\"\n",
    "fold_num = 5\n",
    "labe_fp = \"/home/r10user13/ConcepPath/data/datasets/STAD/molecular_ebv_others.csv\"\n",
    "\n",
    "\n",
    "label_df = pd.read_csv(labe_fp)\n",
    "data_fp_label_map = dict(zip(\n",
    "    list(label_df[\"slide_fp\"]), list(label_df[\"label\"])\n",
    "))\n",
    "\n",
    "exp_rp_ = os.path.join(exp_rp, exp_name, 'input/csv/split')\n",
    "if not os.path.exists(exp_rp_):\n",
    "    os.makedirs(exp_rp_)\n",
    "    \n",
    "\n",
    "if shot_num == \"full_train\":\n",
    "    slide_fp = label_df[\"slide_fp\"]\n",
    "    seg_fp = label_df[\"seg_fp\"]\n",
    "    label = label_df[\"label\"]\n",
    "    \n",
    "    X = np.array(slide_fp)\n",
    "    y = np.array(label)\n",
    "    skf = StratifiedKFold(n_splits=fold_num, shuffle=True, random_state=42)\n",
    "    i = 0 \n",
    "    for train_index, test_index in skf.split(X, y):\n",
    "        X_train, X_test = list(X[train_index]), list(X[test_index])\n",
    "        y_train, y_test = list(y[train_index]), list(y[test_index])\n",
    "        X_val, y_val = X_test, y_test\n",
    "        out = pd.DataFrame({\n",
    "            \"data_path\": X_train+X_test+X_val, \n",
    "            \"label\": y_train+y_test+y_val, \n",
    "            \"type\": [\"train\"]*len(X_train)+[\"test\"]*len(X_test)+[\"val\"]*len(X_val)\n",
    "        })\n",
    "        \n",
    "        out.to_csv(os.path.join(exp_rp_, f\"fold{i}.csv\"))\n",
    "        \n",
    "        i+=1\n",
    "        \n",
    "else:\n",
    "    data_label = {}\n",
    "    label_list = label_df[\"label\"].unique()\n",
    "    \n",
    "    for label_i in label_list:\n",
    "        slide_fp = label_df[label_df[\"label\"]==label_i][\"slide_fp\"]\n",
    "        seg_fp = label_df[label_df[\"label\"]==label_i][\"seg_fp\"]\n",
    "        label = label_df[label_df[\"label\"]==label_i][\"label\"]\n",
    "        \n",
    "        train_data, data_label[label_i][\"test_data\"], data_label[label_i][\"train_label\"], data_label[label_i][\"test_label\"] = train_test_split(\n",
    "            slide_fp, label, test_size=0.2, stratify=label, random_state=2023)\n",
    "        \n",
    "        random.shuffle(train_data)\n",
    "        data_label[label_i][\"train_data\"] = train_data\n",
    "    \n",
    "    for shot in shot_num:\n",
    "        for i in range(fold_num):\n",
    "            \n",
    "            cur_i = i*shot\n",
    "            data_path_list = []\n",
    "            label_list = []\n",
    "            type_list = []\n",
    "            \n",
    "            for label_i in label_list:\n",
    "                \n",
    "                train_data_i = data_label[label_i][\"train_data\"][cur_i:cur_i+shot]\n",
    "                \n",
    "                test_data_i = data_label[label_i][\"test_data\"]\n",
    "                \n",
    "                val_data_i = random.sample(list(set(data_label[label_i][\"train_data\"])-set(train_data_i)), len(test_data_i))\n",
    "                \n",
    "                data_path_list_i = train_data_i+test_data_i+val_data_i\n",
    "                label_list_i = [data_fp_label_map[data_path] for data_path in data_path_list_i]\n",
    "                type_list_i = [\"train\"]*2*shot_num+[\"test\"]*len(test_data_i)+[\"val\"]*len(val_data_i)\n",
    "                \n",
    "                data_path_list += data_path_list_i\n",
    "                label_list += label_list_i\n",
    "                type_list += type_list_i\n",
    "                \n",
    "                \n",
    "            out = pd.DataFrame({\n",
    "                \"data_path\": data_path_list, \n",
    "                \"label\": label_list, \n",
    "                \"type\": type_list\n",
    "            })\n",
    "            out.to_csv(os.path.join(exp_rp_, f\"fold{i}.csv\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Training:\n",
    "\n",
    "For detailed description of parameters, please refer to the *get_params()* function in main.py."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fold0 train\n",
    "!python main.py \\\n",
    "    --is_adapted \\\n",
    "    --weighted_type p2c \\\n",
    "    --orth_ratio 2 \\\n",
    "    --n_ddp 8 \\\n",
    "    --n_ctx 16 \\\n",
    "    --fold_name fold0 \\\n",
    "    --vlm_name quilt1m \\\n",
    "    --feature_rp /data2/r10user13/ConcepPath/stad_quilt1m_20x_448/ \\\n",
    "    --num_epochs 200 \\\n",
    "    --learning_rate 0.0001 \\\n",
    "    --task_type train \\\n",
    "    --n_classes 2 \\\n",
    "    --experiment_rp /home/r10user13/ConcepPath/experiment \\\n",
    "    --experiment_name stad_molecular_ebv_others_re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Testing and saving attention scores:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fold0 test\n",
    "!python main.py \\\n",
    "    --n_flp 8 \\\n",
    "    --n_ctx 16 \\\n",
    "    --model_fp /home/r10user13/ConcepPath/experiment/stad_molecular_ebv_others_ft/output/model/quilt1m_fold0_lr:0.0001_ctx:16_flp:8_specific__test_best_auc_model.pt \\\n",
    "    --fold_name fold0 \\\n",
    "    --vlm_name quilt1m \\\n",
    "    --feature_rp /data2/r10user13/ConcepPath/stad_quilt1m_20x_448/ \\\n",
    "    --task_type test \\\n",
    "    --n_classes 2 \\\n",
    "    --experiment_rp /home/r10user13/ConcepPath/experiment \\\n",
    "    --experiment_name stad_molecular_ebv_others_ft \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Postprocessing:\n",
    "## 5.1 Analyze training metrics\n",
    "\n",
    "Calculate the mean and standard deviation of the metric of 5 folds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "experiment_rp = \"/home/r10user13/ConcepPath/experiment\"\n",
    "experiment_name = \"stad_molecular_ebv_others_re\"\n",
    "\n",
    "import os\n",
    "metric_fp = os.path.join(experiment_rp, experiment_name, \"output/metrics/metrics.csv\")\n",
    "\n",
    "from utils.processing_utils import * \n",
    "\n",
    "metrics_analysis(metric_fp)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5.2 Create heatmap of attention score:\n",
    "\n",
    "Need to run the Step4. \"Testing and saving attention scores\" to create the heatmap after saving the attention score file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# quilt1m\n",
    "!python create_attn_map.py \\\n",
    "    --seg \\\n",
    "    --vlm_model quilt1m \\\n",
    "    --experiment_rp /home/r10user13/ConcepPath/experiment \\\n",
    "    --experiment_name stad_molecular_ebv_others_ft \\\n",
    "    --process_list /home/r10user13/ConcepPath/data/molecular_ebv_others_pl.csv \\\n",
    "    --label_fp /home/r10user13/ConcepPath/data/datasets/STAD/molecular_ebv_others.csv \\\n",
    "    --n_flp 8 \\\n",
    "    --n_classes 2 \\\n",
    "    --n_patch_prompt 26"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 6. Others\n",
    "## 6.1 Prompt References:\n",
    "\n",
    " - HER2 Summary Prompts\n",
    "\n",
    "Shang, Jiuyan et al. “Evolution and clinical significance of HER2-low status after neoadjuvant therapy for breast cancer.” Frontiers in oncology vol. 13 1086480. 22 Feb. 2023, doi:10.3389/fonc.2023.1086480\n",
    "\n",
    "Venetis, Konstantinos et al. “HER2 Low, Ultra-low, and Novel Complementary Biomarkers: Expanding the Spectrum of HER2 Positivity in Breast Cancer.” Frontiers in molecular biosciences vol. 9 834651. 15 Mar. 2022, doi:10.3389/fmolb.2022.834651\n",
    "\n",
    "Zhang, Huina et al. “HER2-low breast cancers: Current insights and future directions.” Seminars in diagnostic pathology vol. 39,5 (2022): 305-312. doi:10.1053/j.semdp.2022.07.003\n",
    "\n",
    "An, Junsha et al. “New Advances in Targeted Therapy of HER2-Negative Breast Cancer.” Frontiers in oncology vol. 12 828438. 4 Mar. 2022, doi:10.3389/fonc.2022.828438\n",
    "\n",
    "Lee, Hyo-Jae et al. “HER2-Positive Breast Cancer: Association of MRI and Clinicopathologic Features With Tumor-Infiltrating Lymphocytes.” AJR. American journal of roentgenology vol. 218,2 (2022): 258-269. doi:10.2214/AJR.21.26400\n",
    "\n",
    "den Hollander, Petra et al. “Targeted therapy for breast cancer prevention.” Frontiers in oncology vol. 3 250. 23 Sep. 2013, doi:10.3389/fonc.2013.00250\n",
    "\n",
    "Patrizio, Armando et al. “Thyroid Metastasis from Primary Breast Cancer.” Journal of clinical medicine vol. 12,7 2709. 4 Apr. 2023, doi:10.3390/jcm12072709\n",
    "\n",
    "\n",
    " - Lung (LUAD vs. LUSC) Summary Prompts\n",
    "\n",
    "Song, Xiaojie et al. “Construction of a Novel Ferroptosis-Related Gene Signature for Predicting Survival of Patients With Lung Adenocarcinoma.” Frontiers in oncology vol. 12 810526. 3 Mar. 2022, doi:10.3389/fonc.2022.810526\n",
    "\n",
    "Qiu, Wang-Ren et al. “Predicting the Lung Adenocarcinoma and Its Biomarkers by Integrating Gene Expression and DNA Methylation Data.” Frontiers in genetics vol. 13 926927. 30 Jun. 2022, doi:10.3389/fgene.2022.926927\n",
    "\n",
    "Wang, Wen et al. “What's the difference between lung adenocarcinoma and lung squamous cell carcinoma? Evidence from a retrospective analysis in a cohort of Chinese patients.” Frontiers in endocrinology vol. 13 947443. 29 Aug. 2022, doi:10.3389/fendo.2022.947443\n",
    "\n",
    "Hu, Xiaoshan et al. “Novel cellular senescence-related risk model identified as the prognostic biomarkers for lung squamous cell carcinoma.” Frontiers in oncology vol. 12 997702. 17 Nov. 2022, doi:10.3389/fonc.2022.997702\n",
    "\n",
    "\n",
    " - Molecular (EBV,MSI,CIN,GS) Summary Prompts\n",
    "\n",
    "Zhu, Chunrong et al. “Genomic Profiling Reveals the Molecular Landscape of Gastrointestinal Tract Cancers in Chinese Patients.” Frontiers in genetics vol. 12 608742. 14 Sep. 2021, doi:10.3389/fgene.2021.608742\n",
    "\n",
    "Dedieu, Stéphane, and Olivier Bouché. “Clinical, Pathological, and Molecular Characteristics in Colorectal Cancer.” Cancers vol. 14,23 5958. 2 Dec. 2022, doi:10.3390/cancers14235958\n",
    "\n",
    "Hinata, Munetoshi, and Tetsuo Ushiku. “Detecting immunotherapy-sensitive subtype in gastric cancer using histologic image-based deep learning.” Scientific reports vol. 11,1 22636. 22 Nov. 2021, doi:10.1038/s41598-021-02168-4\n",
    "\n",
    "Han, Shuting et al. “Epstein-Barr Virus Epithelial Cancers-A Comprehensive Understanding to Drive Novel Therapies.” Frontiers in immunology vol. 12 734293. 10 Dec. 2021, doi:10.3389/fimmu.2021.734293\n",
    "\n",
    "Sun, Keran et al. “EBV-Positive Gastric Cancer: Current Knowledge and Future Perspectives.” Frontiers in oncology vol. 10 583463. 14 Dec. 2020, doi:10.3389/fonc.2020.583463\n",
    "\n",
    "Genitsch, Vera et al. “Epstein-barr virus in gastro-esophageal adenocarcinomas - single center experiences in the context of current literature.” Frontiers in oncology vol. 5 73. 26 Mar. 2015, doi:10.3389/fonc.2015.00073\n",
    "\n",
    "Saito, Motonobu, and Koji Kono. “Landscape of EBV-positive gastric cancer.” Gastric cancer : official journal of the International Gastric Cancer Association and the Japanese Gastric Cancer Association vol. 24,5 (2021): 983-989. doi:10.1007/s10120-021-01215-3\n",
    "\n",
    "Joshi, Smita S, and Brian D Badgwell. “Current treatment and recent progress in gastric cancer.” CA: a cancer journal for clinicians vol. 71,3 (2021): 264-279. doi:10.3322/caac.21657\n",
    "\n",
    "Amato, Martina et al. “Microsatellite Instability: From the Implementation of the Detection to a Prognostic and Predictive Role in Cancers.” International journal of molecular sciences vol. 23,15 8726. 5 Aug. 2022, doi:10.3390/ijms23158726\n",
    "\n",
    "Ratti, Margherita et al. “Microsatellite instability in gastric cancer: molecular bases, clinical perspectives, and new treatment approaches.” Cellular and molecular life sciences : CMLS vol. 75,22 (2018): 4151-4162. doi:10.1007/s00018-018-2906-9\n",
    "\n",
    "Salnikov, Mikhail et al. “Tumor-Infiltrating T Cells in EBV-Associated Gastric Carcinomas Exhibit High Levels of Multiple Markers of Activation, Effector Gene Expression, and Exhaustion.” Viruses vol. 15,1 176. 7 Jan. 2023, doi:10.3390/v15010176\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "clam",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
